### PR TITLE
Add in check for maps added to directory to map directory watcher

### LINKED ISSUE #1879: Maps not added to map list when added to directory manually
Currently maps that are manually added to the maps directory are not added to the maps list in the client forcing a restart for the map to be recognized

**To Reproduce**
Add a map to the maps directory while the client is open

**Expected behavior**
The client should add the map to the installed maps list

**OS**
Windows

### PR DESCRIPTION
**Pull request by Sheikah45**
Fixes #1879

### CODE REVIEW

**Review by 1-alex98 — APPROVED** [review]
Wanna test this still. Maybe we should write a test that validates actions on Create so nobody forgets about it again. Apart from that its nice

**Review by 1-alex98 — CHANGES_REQUESTED** [review]
So I tested it on linux and it seems that installing a map now triggers this as well but it triggers it before it is even fully copied (this would also happen on manually copying if stuff is copied file by file).

The Exception in the log is:
```
2020-09-05 20:55:31.511  WARN 32900 --- [     Thread-101] com.faforever.client.map.MapService      : Map could not be read: adaptive_siberia.v0001

com.faforever.client.map.MapLoadException: Map folder does not contain a *_scenario.lua: /home/alex/.faforever/user/My Games/Gas Powered Games/Supreme Commander Forged Alliance/maps/adaptive_siberia.v0001
	at com.faforever.client.map.MapService.lambda$readMap$10(MapService.java:284)
	at java.base/java.util.Optional.orElseThrow(Optional.java:401)
	at com.faforever.client.map.MapService.readMap(MapService.java:284)
	at com.faforever.client.map.MapService.addInstalledMap(MapService.java:258)
	at com.faforever.client.map.MapService.lambda$startDirectoryWatcher$6(MapService.java:208)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at com.faforever.client.map.MapService.lambda$startDirectoryWatcher$7(MapService.java:204)
	at com.github.nocatch.NoCatch.noCatch(NoCatch.java:49)
	at com.github.nocatch.NoCatch.noCatch(NoCatch.java:34)
	at com.faforever.client.map.MapService.lambda$startDirectoryWatcher$8(MapService.java:197)
	at java.base/java.lang.Thread.run(Thread.java:832)
```
Such a *_scenario.lua file exists but is not yet copied. We need to do something to avoid this. Either we have a look that needs to be acquired or we just delay the watcher by a few seconds so it does not trigger in between copying files.

**Comment by Sheikah45** [comment]
Well if we delay should it be delayed by longer for various download speeds?

**Comment by Sheikah45** [comment]
I added in a watcher for when the scenario file is written and gave it a timeout of a minute so this should cover the situations of downloading the map or manually copying them.

**Comment by 1-alex98** [comment]
@Sheikah45 The map is first downloaded and then unpacked. So a delay of a second would be enough

**Comment by 1-alex98** [comment]
OK the whole delay thing is probably shit. I looked at the methods that add a map and it only reads from the scenario.lua. So let's drop all the delay thing and just react to scenario.lua being added and drop all other create actions

**Comment by 1-alex98** [comment]
ok we go for sub directory watcher right?

**Comment by Sheikah45** [comment]
Yes and I am going to move it into a function

**Comment by Sheikah45** [comment]
Alright moved it to a sub function. Let me know if you want anything changed.

**Comment by Sheikah45** [comment]
Hmm CI is failing but all tests pass when I run them

File: src/main/java/com/faforever/client/map/MapService.java [hunk 1]
```diff
@@ -266,6 +224,52 @@ private Thread startDirectoryWatcher(Path mapsDirectory) {
     return thread;
   }
 
+  private boolean checkScenarioLua(Path mapPath) {
```
Comment (1-alex98) [inline src/main/java/com/faforever/client/map/MapService.java]: Normally we don't make comments but as this is not a common thing to do maybe explain it in a java doc

**Comment by Sheikah45** [comment]
are there examples of how you think it should be documented

File: src/main/java/com/faforever/client/map/MapService.java (hunk 1 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/map/MapService.java]: are there examples of how you think it should be documented

Comment (Sheikah45) [reply src/main/java/com/faforever/client/map/MapService.java]: Added in the javadoc

**Review by Brutus5000 — COMMENTED** [review]
Never have I seen such a horrible combination of blocking and threading.

I see a huge load of complexity just for the case that the user could have create a directory that doesn't really contain a map. Wouldn't it be easier to say: Hey, we waited 5 seconds still no scenaria.lua we throw an error?

Alternatively I would think about using a proper async library like reactor which gives you retry *and* error handling. (But this is a general thing to be discussed I guess, we shouldn'T just introduce it for this case)

**Comment by Sheikah45** [comment]
Yeah we can just set it to a timer I just wasn't sure if there would be edge cases that should work but fail with just a timer

**Comment by 1-alex98** [comment]
Yes we know that it is suboptimal. But the standard java lib offers to better it seems(at least not if it should work on linux and windows). Tell us what to do @Brutus5000 XD

**Comment by Sheikah45** [comment]
I just used a second delay

**Comment by 1-alex98** [comment]
@Brutus5000 u ok with this?

**Comment by Sheikah45** [comment]
anything to do here?

**Comment by 1-alex98** [comment]
@Brutus5000 2 days for your opinion other wise gonna merge

**Review by Brutus5000 — DISMISSED** [review]

**Comment by 1-alex98** [comment]
Seems like I fucked up

**Comment by Sheikah45** [comment]
why close? Do I need to change something?

**Comment by Sheikah45** [comment]
oh wait was it merged?

**Pull request closed without merging**